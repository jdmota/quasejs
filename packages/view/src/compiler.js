import TreeWalker from "./tree-walker";

const parse5 = require( "parse5" );

/* eslint-disable no-control-regex */
const lastAttrNameRegex =
  /[ \x09\x0a\x0c\x0d]([^\0-\x1F\x7F-\x9F \x09\x0a\x0c\x0d"'>=/]+)[ \x09\x0a\x0c\x0d]*=[ \x09\x0a\x0c\x0d]*(?:[^ \x09\x0a\x0c\x0d"'`<>=]*|"[^"]*|'[^']*)$/;
const marker = `{{quase-view-${String( Math.random() ).slice( 2 )}}}`;
const nodeMarker = `<!--${marker}-->`;

function findTagClose( str ) {
  const close = str.lastIndexOf( ">" );
  const open = str.indexOf( "<", close + 1 );
  return open > -1 ? str.length : close;
}

function getHtml( strings, svg ) {
  const l = strings.length - 1;
  let html = "";
  let isTextBinding = true;
  for ( let i = 0; i < l; i++ ) {
    const s = strings[ i ];
    html += s;
    // If the previous string closed its tags, we're in a text position.
    // If it doesn't have any tags, then we use the previous text position
    // state.
    const closing = findTagClose( s );
    isTextBinding = closing > -1 ? closing < s.length : isTextBinding;
    html += isTextBinding ? nodeMarker : marker;
  }
  html += strings[ l ];
  return svg ? `<svg>${html}</svg>` : html;
}

const treeAdapter = parse5.treeAdapters.htmlparser2;

const reOn = /^on-/i;

function handleAttr( t, index, attrName, attrNameInPart, strings ) {
  if ( reOn.test( attrName ) ) {
    return t.arrayExpression(
      [
        t.stringLiteral( "event" ),
        t.numericLiteral( index ),
        t.stringLiteral( attrNameInPart.slice( 3 ) )
      ]
    );
  }
  if ( attrNameInPart.endsWith( "$" ) ) {
    return t.arrayExpression(
      [
        t.stringLiteral( "attr" ),
        t.numericLiteral( index ),
        t.stringLiteral( attrNameInPart.slice( 0, -1 ) ),
        t.arrayExpression( strings.map( s => t.stringLiteral( s ) ) )
      ]
    );
  }
  return t.arrayExpression(
    [
      t.stringLiteral( "prop" ),
      t.numericLiteral( index ),
      t.stringLiteral( attrNameInPart ),
      t.arrayExpression( strings.map( s => t.stringLiteral( s ) ) )
    ]
  );
}

function handleNode( t, index ) {
  return t.arrayExpression( [ t.stringLiteral( "node" ), t.numericLiteral( index ) ] );
}

export default function compile( strings, svg, t ) {

  const code = getHtml( strings, svg );
  const doc = parse5.parseFragment( code, { treeAdapter } );
  const parts = [];
  const walker = new TreeWalker( doc );

  let index = 0;
  let partIndex = 0;

  while ( walker.nextNode() ) {
    const node = walker.currentNode;

    if ( treeAdapter.isElementNode( node ) ) {

      for ( const attrName in node.attribs ) {
        const value = node.attribs[ attrName ];
        if ( value.indexOf( marker ) >= 0 ) {
          const attrNameInPart = lastAttrNameRegex.exec( strings[ partIndex ] )[ 1 ];
          const stringsForAttr = value.split( marker );
          delete node.attribs[ attrName ];
          parts.push( handleAttr( t, index, attrName, attrNameInPart, stringsForAttr ) );
          partIndex += stringsForAttr.length - 1;
        }
      }

    } else if ( treeAdapter.isCommentNode( node ) && node.data === marker ) {

      node.data = "{{}}";
      parts.push( handleNode( t, index ) );
      partIndex++;

    }

    index++;
  }

  const html = parse5.serialize( doc, { treeAdapter } );

  return {
    parts,
    html
  };
}

export function babelPlugin( { types: t } ) {
  return {
    visitor: {
      TaggedTemplateExpression( path ) {
        if ( path.node.tag.name !== "html" && path.node.tag.name !== "svg" ) {
          return;
        }
        const quaseViewTemplate = t.memberExpression( t.identifier( "QuaseView" ), t.identifier( "Template" ) );
        const quaseViewTemplateResult = t.memberExpression( t.identifier( "QuaseView" ), t.identifier( "TemplateResult" ) );

        const { parts, html } = compile(
          path.node.quasi.quasis.map( q => q.value.cooked ),
          path.node.tag.name === "svg",
          t
        );

        path.replaceWith(
          t.newExpression(
            quaseViewTemplateResult,
            [
              t.newExpression(
                quaseViewTemplate,
                [ t.arrayExpression( parts ), t.stringLiteral( html ) ]
              ),
              t.arrayExpression( path.node.quasi.expressions )
            ]
          )
        );
      }
    }
  };
}
