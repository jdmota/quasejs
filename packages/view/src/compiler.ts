// @ts-ignore
import jsx from "@babel/plugin-syntax-jsx";
import {
  PART_ATTR, PART_PROP, PART_EVENT, PART_NODE,
  TEMPLATE, TEMPLATE_RESULT,
  partValue
} from "./spec";

type BabelTypes = any;
type BabelPath = any;

type Expression = {
  type: "unknown";
  [key: string]: unknown;
};

type NumericLiteral = {
  type: "NumericLiteral";
  value: number;
};

type StringLiteral = {
  type: "StringLiteral";
  value: string;
  extra: {
    rawValue: string;
    raw: string;
  };
};

type JSXEmptyExpression = {
  type: "JSXEmptyExpression";
};

type JSXIdentifier = {
  type: "JSXIdentifier";
  name: string;
};

type JSXText = {
  type: "JSXText";
  value: string;
  extra: {
    rawValue: string;
    raw: string;
  };
};

type JSXExpressionContainer = {
  type: "JSXExpressionContainer";
  expression: JSXEmptyExpression | StringLiteral | Expression;
};

type JSXAttribute = {
  type: "JSXAttribute";
  name: JSXIdentifier;
  value: StringLiteral | JSXExpressionContainer;
};

type JSXOpeningElement = {
  type: "JSXOpeningElement";
  name: JSXIdentifier;
  attributes: JSXAttribute[];
  selfClosing: boolean;
};

type JSXElement = {
  type: "JSXElement";
  openingElement: JSXOpeningElement;
  children: JSXChildren[];
};

type JSXFragment = {
  type: "JSXFragment";
  children: JSXChildren[];
};

type JSXChildren = JSXElement | JSXFragment | JSXText | JSXExpressionContainer;

const escapeHTML = require( "he" ).encode;
const reOn = /^on[A-Z]/;

function handleAttr( result: Result, t: BabelTypes, firstAttr: boolean, attrName: string ) {
  let type: number;
  let string: string;
  if ( reOn.test( attrName ) ) {
    type = PART_EVENT;
    string = attrName.slice( 2 ).toLowerCase();
  } else if ( attrName.endsWith( "$" ) ) {
    type = PART_ATTR;
    string = attrName.slice( 0, -1 );
  } else {
    type = PART_PROP;
    string = attrName;
  }
  result.parts.push( t.numericLiteral( partValue( firstAttr, type ) ) );
  result.strings.push( t.stringLiteral( string ) );
}

function handleNode( result: Result, t: BabelTypes ) {
  result.parts.push(
    t.numericLiteral( partValue( true, PART_NODE ) )
  );
}

const visitor = {
  JSXElement( result: Result, t: BabelTypes, { openingElement, children }: JSXElement ) {
    const hasParts = visitor.hasAttrParts( openingElement );
    if ( hasParts ) {
      result.html.push( `<!---->` );
    }

    if ( openingElement.selfClosing ) {
      result.html.push( `<${openingElement.name.name}` );
    } else {
      result.html.push( `<${openingElement.name.name}>` );
    }

    let firstAttr = true;
    for ( const attr of openingElement.attributes ) {
      visitor.JSXAttribute( result, t, attr, firstAttr );
      firstAttr = false;
    }

    if ( openingElement.selfClosing ) {
      result.html.push( `/>` );
    } else {

      for ( const child of children ) {
        switch ( child.type ) {
          case "JSXElement":
            visitor.JSXElement( result, t, child );
            break;
          case "JSXFragment":
            visitor.JSXFragment( result, t, child );
            break;
          case "JSXText":
            result.html.push( escapeHTML( child.value ) );
            break;
          case "JSXExpressionContainer":
            visitor.JSXExpressionContainer( result, t, child );
            break;
          default:
            throw new Error( `Unknown child of type ${( child as any ).type}` );
        }
      }

      result.html.push( `</${openingElement.name.name}>` );
    }

  },
  JSXFragment( result: Result, t: BabelTypes, { children }: JSXFragment ) {

    for ( const child of children ) {
      switch ( child.type ) {
        case "JSXElement":
          visitor.JSXElement( result, t, child );
          break;
        case "JSXFragment":
          visitor.JSXFragment( result, t, child );
          break;
        case "JSXText":
          result.html.push( escapeHTML( child.value ) );
          break;
        case "JSXExpressionContainer":
          visitor.JSXExpressionContainer( result, t, child );
          break;
        default:
          throw new Error( `Unknown child of type ${( child as any ).type}` );
      }
    }

  },
  hasAttrParts( openingElement: JSXOpeningElement ) {
    for ( const { value } of openingElement.attributes ) {
      switch ( value.type ) {
        case "StringLiteral":
          break;
        case "JSXExpressionContainer": {
          const { expression } = value;
          switch ( expression.type ) {
            case "JSXEmptyExpression":
              throw new Error( `Cannot use JSXEmptyExpression in JSXAttribute` );
            case "StringLiteral":
              break;
            default:
              return true;
          }
          break;
        }
        default:
          throw new Error( `Unknown attr value of type ${( value as any ).type}` );
      }
    }
    return false;
  },
  JSXAttribute( result: Result, t: BabelTypes, { name: { name }, value }: JSXAttribute, firstAttr: boolean ) {
    switch ( value.type ) {
      case "StringLiteral":
        result.html.push( `${name}="${escapeHTML( value.value )}"` );
        break;
      case "JSXExpressionContainer": {
        const { expression } = value;
        switch ( expression.type ) {
          case "JSXEmptyExpression":
            throw new Error( `Cannot use JSXEmptyExpression in JSXAttribute` );
          case "StringLiteral":
            result.html.push( `${name}="${escapeHTML( expression.value )}"` );
            break;
          default:
            handleAttr( result, t, firstAttr, name );
            result.expressions.push( expression );
        }
        break;
      }
      default:
        throw new Error( `Unknown attr value of type ${( value as any ).type}` );
    }
  },
  JSXExpressionContainer( result: Result, t: BabelTypes, { expression }: JSXExpressionContainer ) {
    switch ( expression.type ) {
      case "JSXEmptyExpression":
        break;
      case "StringLiteral":
        result.html.push( escapeHTML( expression.value ) );
        break;
      default:
        handleNode( result, t );
        result.html.push( "<!---->" );
        result.expressions.push( expression );
    }
  }
};

type Result = {
  html: string[];
  parts: NumericLiteral[];
  strings: StringLiteral[];
  expressions: Expression[];
};

type Fn =
  ( ( r: Result, t: BabelTypes, n: JSXElement ) => void ) |
  ( ( r: Result, t: BabelTypes, n: JSXFragment ) => void );

function mainVisitor(
  t: BabelTypes,
  fn: Fn,
  path: BabelPath
) {

  const { node } = path;

  const quaseViewTemplate = t.memberExpression( t.identifier( "QuaseView" ), t.identifier( TEMPLATE ) );
  const quaseViewTemplateResult = t.memberExpression( t.identifier( "QuaseView" ), t.identifier( TEMPLATE_RESULT ) );

  const result: Result = {
    html: [],
    parts: [],
    strings: [],
    expressions: []
  };

  fn( result, t, node );

  const program = path.scope.getProgramParent().path;

  const templateUID = program.scope.generateUidIdentifier( "render" );

  const templateDeclaration = t.variableDeclaration( "const", [
    t.variableDeclarator(
      templateUID,
      t.newExpression(
        quaseViewTemplate,
        [
          t.arrayExpression( result.parts ),
          t.arrayExpression( result.strings ),
          t.stringLiteral( result.html.join( "" ) )
        ]
      )
    )
  ] );

  program.unshiftContainer( "body", templateDeclaration );

  path.replaceWith(
    t.newExpression(
      quaseViewTemplateResult,
      [
        templateUID,
        t.arrayExpression( result.expressions )
      ]
    )
  );

}

export default function( { types: t }: { types: BabelTypes } ) {
  return {
    inherits: jsx,
    visitor: {
      JSXElement: ( path: BabelPath ) => mainVisitor( t, visitor.JSXElement, path ),
      JSXFragment: ( path: BabelPath ) => mainVisitor( t, visitor.JSXFragment, path )
    }
  };
}
