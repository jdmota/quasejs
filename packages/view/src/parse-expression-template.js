// Adapted from https://github.com/Polymer/polymer/blob/master/lib/mixins/property-effects.html
const bindingRegex = ( function() {
  const IDENT = "(?:[a-zA-Z_$][\\w.:$\\-*]*)";
  const NUMBER = "(?:[-+]?[0-9]*\\.?[0-9]+(?:[eE][-+]?[0-9]+)?)";
  const SQUOTE_STRING = "(?:'(?:[^'\\\\]|\\\\.)*')";
  const DQUOTE_STRING = "(?:\"(?:[^\"\\\\]|\\\\.)*\")";
  const STRING = "(?:" + SQUOTE_STRING + "|" + DQUOTE_STRING + ")";
  const ARGUMENT = "(?:" + IDENT + "|" + NUMBER + "|" + STRING + "\\s*)";
  const ARGUMENTS = "(?:" + ARGUMENT + "(?:,\\s*" + ARGUMENT + ")*)";
  const ARGUMENT_LIST = "(?:\\(\\s*(?:" + ARGUMENTS + "?)\\)\\s*)";
  const BINDING = "((?:!\\s*)?" + IDENT + "\\s*" + ARGUMENT_LIST + "?)"; // Group 1
  const OPEN_BRACKET = "(?:\\[\\[|{{)\\s*";
  const CLOSE_BRACKET = "(?:]]|}})";
  return new RegExp( OPEN_BRACKET + BINDING + CLOSE_BRACKET, "g" );
} )();

// "literal1{{prop}}literal2{{!compute(foo,bar)}}final"
// [ "literal1", "prop", "literal2", "!compute(foo,bar)", "final" ]
export default function parseExpressionTemplate( text ) {

  let parts = [], lastIndex = 0, m;

  while ( m = bindingRegex.exec( text ) ) {
    parts.push( m.index > lastIndex ? text.slice( lastIndex, m.index ) : "" );
    parts.push( m[ 1 ].trim() );
    lastIndex = bindingRegex.lastIndex;
  }

  parts.push( text.substring( lastIndex ) );

  return parts;
}

/*

const babylon = req( "babylon" );

function parseJs( text, location ) {
  try {
    return babylon.parseExpression( text, { strictMode: true } );
  } catch ( e ) {
    error( "Invalid expression", location );
  }
}

*/
