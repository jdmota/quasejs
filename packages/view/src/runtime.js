( function( factory ) {
  if ( typeof module !== "undefined" && module.exports ) {
    module.exports = factory;
  } else {
    factory( window ); // eslint-disable-line no-undef
  }
} )( function( win ) {

  const doc = win.document;

  function handlePart( instance, node, part ) {
    switch ( part[ 0 ] ) {
      case "attr":
        return new AttributePart( instance, node, part[ 2 ], part[ 3 ] );
      case "prop":
        return new PropertyPart( instance, node, part[ 2 ], part[ 3 ] );
      case "event":
        return new EventPart( instance, node, part[ 2 ] );
      case "node":
        return new NodePart( instance, node );
      default:
    }
    throw new Error( `Unknown part type ${part[ 0 ]}` );
  }

  class Template {
    constructor( parts, html ) {
      this.element = doc.createElement( "template" );
      this.element.innerHTML = html;
      this.parts = parts;
    }
  }

  class TemplateResult {
    constructor( template, values ) {
      this.template = template;
      this.values = values;
    }
  }

  class TemplateInstance {

    constructor( template ) {
      this.template = template;
      this.parts = [];
    }

    update( values ) {
      let valueIndex = 0;
      for ( const part of this.parts ) {
        if ( part.size === undefined ) {
          part.setValue( values[ valueIndex ] );
          valueIndex++;
        } else {
          part.setValue( values, valueIndex );
          valueIndex += part.size;
        }
      }
    }

    createNode() {
      const fragment = doc.importNode( this.template.element.content, true );
      const parts = this.template.parts;

      if ( parts.length > 0 ) {
        // Edge needs all 4 parameters present
        const walker = doc.createTreeWalker(
          fragment,
          129, /* NodeFilter.SHOW_ELEMENT | NodeFilter.SHOW_COMMENT */
          null,
          false
        );

        let index = -1;
        for ( let i = 0; i < parts.length; i++ ) {
          const part = parts[ i ];
          while ( index < part[ 1 ] ) {
            index++;
            walker.nextNode();
          }
          this.parts.push( handlePart( this, walker.currentNode, part ) );
        }
      }
      return fragment;
    }
  }

  class Part {
    constructor( instance, element ) {
      this.instance = instance;
      this.element = element;
    }
    getValue( value ) {
      if ( typeof value === "function" && value.__directive ) {
        value = value( this );
      }
      return value == null ? null : value;
    }
    getValueNoFalse( value ) {
      const v = this.getValue( value );
      return v === false ? null : v;
    }
  }

  class AttributePart extends Part {

    constructor( instance, element, name, strings ) {
      super( instance, element );
      this.name = name;
      this.strings = strings;
      this.size = strings.length - 1;
      this._previousValue = null;
    }

    _interpolate( values, startIdx ) {
      const s = this.strings;
      if ( this.size === 1 && s[ 0 ] === "" && s[ 1 ] === "" ) {
        return this.getValueNoFalse( values[ startIdx ] );
      }

      let text = "";
      for ( let i = 0; i < this.size; i++ ) {
        text += s[ i ] + this.getValueNoFalse( values[ startIdx + i ] );
      }
      return text + s[ this.size ];
    }

    setValue( values, startIdx ) {
      const value = this._interpolate( values, startIdx );
      if ( this._previousValue !== value ) {
        if ( value == null ) {
          this.element.removeAttribute( this.name );
        } else {
          this.element.setAttribute( this.name, value );
        }
        this._previousValue = value;
      }
    }
  }

  class PropertyPart extends AttributePart {
    setValue( values, startIdx ) {
      this.element[ this.name ] = this._interpolate( values, startIdx );
    }
  }

  class EventPart extends Part {

    constructor( instance, element, eventName ) {
      super( instance, element );
      this.eventName = eventName;
      this._listener = null;
    }

    setValue( value ) {
      const listener = this.getValueNoFalse( value );
      const previous = this._listener;
      if ( listener === previous ) {
        return;
      }

      this._listener = listener;
      if ( previous != null ) {
        this.element.removeEventListener( this.eventName, previous );
      }
      if ( listener != null ) {
        this.element.addEventListener( this.eventName, listener );
      }
    }
  }

  class NodePart extends Part {

    constructor( instance, element ) {
      super( instance, element );
      this._previousValue = undefined;
    }

    setValue( value ) {
      value = this.getValue( value );

      if ( value == null || !( typeof value === "object" || typeof value === "function" ) ) {
        if ( value === this._previousValue ) {
          return;
        }
        this._setText( value );
      } else if ( value instanceof TemplateResult ) {
        this._setTemplateResult( value );
      } else if ( Array.isArray( value ) || value[ Symbol.iterator ] ) {
        this._setIterable( value );
      } else if ( value instanceof win.Node ) {
        this._setNode( value );
      } else if ( value.then != null ) {
        this._setPromise( value );
      } else {
        this._setText( value );
      }
    }

    _replace( node ) {
      this.element.parentNode.replaceChild( node, this.element );
      this.element = node;
    }

    _setNode( value ) {
      if ( this._previousValue === value ) {
        return;
      }
      this._replace( value );
      this._previousValue = value;
    }

    _setText( value ) {
      if ( this.element.nodeType === 3 /* Node.TEXT_NODE */ ) {
        this.element.textContent = value;
      } else {
        this._setNode( doc.createTextNode( value == null ? "" : value ) );
      }
      this._previousValue = value;
    }

    _setTemplateResult( value ) {
      let instance;
      if ( this._previousValue && this._previousValue.template === value.template ) {
        instance = this._previousValue;
      } else {
        instance = new TemplateInstance( value.template );
        this._setNode( instance.createNode() );
        this._previousValue = instance;
      }
      instance.update( value.values );
    }

    _setPromise( value ) {
      this._previousValue = value;
      value.then( v => {
        if ( this._previousValue === value ) {
          this.setValue( v );
        }
      } );
    }

    _setIterable( value ) {
      if ( !Array.isArray( this._previousValue ) ) {
        this._replace( doc.createComment( "" ) );
        this._previousValue = [];
      }

      const itemParts = this._previousValue;
      let partIndex = 0;

      for ( const item of value ) {
        let itemPart = itemParts[ partIndex ];

        if ( partIndex >= itemParts.length ) {
          itemPart = new NodePart( this.instance, doc.createComment( "" ) );
          itemParts.push( itemPart );
          this.element.parentNode.insertBefore( itemPart.element, this.element );
        }

        itemPart.setValue( item );
        partIndex++;
      }

      for ( let i = partIndex; i < itemParts.length; i++ ) {
        const element = itemParts[ i ].element;
        element.parentNode.removeChild( element );
      }
      itemParts.length = partIndex;
    }

  }

  function render( result, container ) {
    let instance = container.__templateInstance;

    if ( instance != null && instance.template === result.template ) {
      instance.update( result.values );
      return;
    }

    instance = new TemplateInstance( result.template );
    container.__templateInstance = instance;

    const fragment = instance.createNode();
    instance.update( result.values );

    container.appendChild( fragment );
  }

  class QuaseView extends win.HTMLElement {
    constructor() {
      super();
      this._ready = false;
      this.attachShadow( { mode: "open" } );
    }
    static get observedAttributes() {
      // Note: `this` is the class itself
      return Object.keys( this.props || {} );
    }
    forceUpdate() {
      QuaseView.render( this.render(), this.shadowRoot );
    }
    connectedCallback() {
      this._ready = true;
      this.forceUpdate();
    }
    attributeChangedCallback( name, oldValue, newValue ) {
      if ( this._ready && oldValue !== newValue ) {
        this.forceUpdate();
      }
    }
  }

  QuaseView.Template = Template;
  QuaseView.TemplateResult = TemplateResult;
  QuaseView.render = render;

  const modules = Object.create( null );

  QuaseView.define = function( name, clazz ) {
    if ( modules[ name ] ) {
      throw new Error( "QuaseView: module " + name + " already defined" );
    }
    modules[ name ] = true;
    win.customElements.define( name, clazz );
  };

  win.QuaseView = QuaseView;
  return QuaseView;
} );
