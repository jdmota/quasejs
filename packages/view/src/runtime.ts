// Inspired in https://github.com/Polymer/lit-html/

type Win = Window & {
  HTMLElement: typeof HTMLElement; // eslint-disable-line no-undef
  Node: typeof Node; // eslint-disable-line no-undef
};

(function (factory) {
  if (typeof module !== "undefined" && module.exports) {
    module.exports = factory;
  } else {
    factory(window as Win); // eslint-disable-line no-undef
  }
})(function (win: Win) {
  const doc = win.document;

  class Template {
    element: HTMLTemplateElement;
    parts: number[];
    strings: string[];

    constructor(parts: number[], strings: string[], html: string) {
      this.element = doc.createElement("template");
      this.element.innerHTML = html;
      this.parts = parts;
      this.strings = strings;
    }
  }

  class TemplateResult {
    template: Template;
    values: unknown[];

    constructor(template: Template, values: unknown[]) {
      this.template = template;
      this.values = values;
    }
  }

  function malformedTemplate(): never {
    throw new Error("Quase View: malformed template");
  }

  // Edge needs all 4 parameters present
  const walker = doc.createTreeWalker(
    doc,
    128, // NodeFilter.SHOW_COMMENT
    null
  );

  const getWalker = (root: Node): TreeWalker => {
    if (walker.currentNode !== doc) {
      throw new Error("Quase View: walker was not reset");
    }
    walker.currentNode = root;
    return walker;
  };

  // TreeWalker should not hold onto any GC'able Nodes
  const resetWalker = () => {
    walker.currentNode = doc;
  };

  class TemplateInstance {
    template: Template;
    parts: Part[];

    constructor(template: Template) {
      this.template = template;
      this.parts = [];
    }

    update(values: unknown[]) {
      let i = 0;
      for (const part of this.parts) {
        part.setValue(values[i++]);
      }
    }

    createNode() {
      const fragment = this.template.element.content.cloneNode(true);
      const parts = this.template.parts;
      const strings = this.template.strings;

      if (parts.length > 0) {
        const walker = getWalker(fragment);

        let strIdx = 0;
        let mark = null;
        let node = null;

        for (let i = 0; i < parts.length; i++) {
          const partInfo = parts[i];
          const newMark = partInfo >>> 2 !== 0;
          const partType = partInfo & 0b11;

          if (newMark) {
            mark = walker.nextNode() || malformedTemplate();
            node = mark.nextSibling;
            if (partType < 3) {
              (mark as ChildNode).remove();
              if (node) {
                walker.currentNode = node;
              }
            }
          }

          let part: Part;

          switch (partType) {
            case 0:
              part = new AttributePart(
                (node || malformedTemplate()) as Element,
                strings[strIdx++]
              );
              break;
            case 1:
              part = new PropertyPart(
                (node || malformedTemplate()) as Element,
                strings[strIdx++]
              );
              break;
            case 2:
              part = new EventPart(
                node || malformedTemplate(),
                strings[strIdx++]
              );
              break;
            case 3:
              part = new NodePart(mark || malformedTemplate());
              break;
            default:
              throw malformedTemplate();
          }

          this.parts.push(part);
        }

        resetWalker();
      }
      return fragment;
    }
  }

  abstract class Part {
    protected element: Node;
    protected _value: any;

    constructor(element: Node) {
      this.element = element;
      this._value = undefined;
    }
    getValue(value: any) {
      if (typeof value === "function" && value.__directive) {
        value = value(this);
      }
      return value == null ? null : value;
    }
    getValueNoFalse(value: any) {
      const v = this.getValue(value);
      return v === false ? null : v;
    }
    abstract setValue(value: any): void;
  }

  class AttributePart extends Part {
    protected name: string;
    protected override element: Element;

    constructor(element: Element, name: string) {
      super(element);
      this.name = name;
      this.element = element;
    }

    setValue(_value: any) {
      const value = this.getValueNoFalse(_value);
      if (this._value !== value) {
        if (value == null) {
          this.element.removeAttribute(this.name);
        } else {
          this.element.setAttribute(this.name, value);
        }
        this._value = value;
      }
    }
  }

  class PropertyPart extends AttributePart {
    override setValue(value: any) {
      (this.element as any)[this.name] = value;
    }
  }

  class EventPart extends Part {
    protected eventName: string;

    constructor(element: Node, eventName: string) {
      super(element);
      this.eventName = eventName;
    }

    setValue(value: any) {
      const listener = this.getValueNoFalse(value);
      const previous = this._value;
      if (listener === previous) {
        return;
      }

      this._value = listener;
      if (previous != null) {
        this.element.removeEventListener(this.eventName, previous);
      }
      if (listener != null) {
        this.element.addEventListener(this.eventName, listener);
      }
    }
  }

  class NodePart extends Part {
    setValue(value: any) {
      value = this.getValue(value);

      if (
        value == null ||
        !(typeof value === "object" || typeof value === "function")
      ) {
        if (value === this._value) {
          return;
        }
        this._setText(value);
      } else if (value instanceof TemplateResult) {
        this._setTemplateResult(value);
      } else if (value instanceof win.Node) {
        this._setNode(value);
      } else if (value[Symbol.iterator]) {
        this._setIterable(value);
      } else if (value.then != null) {
        this._setPromise(value);
      } else {
        this._setText(value);
      }
    }

    private _replace(node: any) {
      if (this.element.parentNode) {
        this.element.parentNode.replaceChild(node, this.element);
      } else {
        throw new Error("Assertion error");
      }
      this.element = node;
    }

    private _setNode(value: any) {
      if (this._value === value) {
        return;
      }
      this._replace(value);
      this._value = value;
    }

    private _setText(value: string) {
      if (this.element.nodeType === 3 /* Node.TEXT_NODE */) {
        this.element.textContent = value;
      } else {
        this._setNode(doc.createTextNode(value == null ? "" : value));
      }
      this._value = value;
    }

    private _setTemplateResult(value: TemplateResult) {
      let instance: TemplateInstance;
      if (this._value && this._value.template === value.template) {
        instance = this._value;
      } else {
        instance = new TemplateInstance(value.template);
        this._setNode(instance.createNode());
        this._value = instance;
      }
      instance.update(value.values);
    }

    private _setPromise(value: Promise<any>) {
      this._value = value;
      value.then(v => {
        if (this._value === value) {
          this.setValue(v);
        }
      });
    }

    private _setIterable(value: Iterable<any>) {
      if (!this._value || !this._value[Symbol.iterator]) {
        this._replace(doc.createComment(""));
        this._value = [];
      }

      if (!this.element.parentNode) {
        throw new Error("Assertion error");
      }

      const itemParts = this._value;
      let partIndex = 0;

      for (const item of value) {
        let itemPart = itemParts[partIndex];

        if (partIndex >= itemParts.length) {
          itemPart = new NodePart(doc.createComment(""));
          itemParts.push(itemPart);
          this.element.parentNode.insertBefore(itemPart.element, this.element);
        }

        itemPart.setValue(item);
        partIndex++;
      }

      for (let i = partIndex; i < itemParts.length; i++) {
        const element = itemParts[i].element;
        element.parentNode.removeChild(element);
      }
      itemParts.length = partIndex;
    }
  }

  const nodeToInstance: WeakMap<Node, TemplateInstance> = new WeakMap();

  function render(result: TemplateResult, container: Node) {
    let instance = nodeToInstance.get(container);

    if (instance != null && instance.template === result.template) {
      instance.update(result.values);
      return;
    }

    instance = new TemplateInstance(result.template);
    nodeToInstance.set(container, instance);

    const fragment = instance.createNode();
    container.appendChild(fragment);

    // Update after custom elements upgrade
    instance.update(result.values);
  }

  const modules = Object.create(null);

  function define(name: string, clazz: CustomElementConstructor) {
    if (modules[name]) {
      throw new Error("QuaseView: module " + name + " already defined");
    }
    modules[name] = true;
    win.customElements.define(name, clazz);
  }

  abstract class QuaseView extends win.HTMLElement {
    static t = Template;
    static r = TemplateResult;
    static render = render;
    static define = define;
    static props: any;

    private _ready: boolean;

    constructor() {
      super();
      this._ready = false;
      this.attachShadow({ mode: "open" });
    }
    static get observedAttributes() {
      // Note: `this` is the class itself
      return Object.keys(this.props || {});
    }
    forceUpdate() {
      QuaseView.render(this.render(), this.shadowRoot as ShadowRoot);
    }
    connectedCallback() {
      this._ready = true;
      this.forceUpdate();
    }
    attributeChangedCallback(_name: string, oldValue: any, newValue: any) {
      if (this._ready && oldValue !== newValue) {
        this.forceUpdate();
      }
    }
    render(): TemplateResult {
      throw new Error("QuaseView: unimplemented render method");
    }
  }

  (win as any).QuaseView = QuaseView;
  return QuaseView;
});

export {};
