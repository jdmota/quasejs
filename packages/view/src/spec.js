export const PART_ATTR = 0;
export const PART_PROP = 1;
export const PART_EVENT = 2;
export const PART_NODE = 3;

export const TEMPLATE = "t";
export const TEMPLATE_RESULT = "r";
