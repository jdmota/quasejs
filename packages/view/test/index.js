import { babelPlugin } from "../src/compiler";

const babel = require( "@babel/core" );
const jsdom = require( "jsdom" );
const { JSDOM } = jsdom;

const fixtures = [
  `
  render = function render() {
    return html\`
    <div
      someProp="\${1}"
      a-nother="\${2}"
      multiParts='\${3} \${4}'
      👍=\${5}
      (a)=\${6}
      [a]=\${7}
      a$=\${8}>
      <p>\${9}</p>
      <div aThing="\${10}">\${[1,html\`<span>\${2}</span>\`,3]}</div>
    </div>\`
  };
  `
];

describe( "compile and runtime", () => {

  let i = 0;

  for ( const code of fixtures ) {
    it( `test ${i++}`, () => {

      const finalCode = babel.transform( code, {
        plugins: [ babelPlugin ],
        sourceMaps: false,
        generatorOpts: {
          retainLines: true
        }
      } ).code;

      expect( finalCode ).toMatchSnapshot();

      const dom = new JSDOM( `<!DOCTYPE html><div id="container"></div>` );
      const container = dom.window.document.getElementById( "container" );
      const QuaseView = require( "../src/runtime" )( dom.window );

      let render;
      eval( finalCode ); // eslint-disable-line no-eval
      QuaseView.render( render(), container );

      expect( container.innerHTML ).toMatchSnapshot();

    } );
  }

} );
