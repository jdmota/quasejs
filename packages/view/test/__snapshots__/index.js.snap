// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile and runtime test 0 1`] = `
"
render = function render() {
  return new QuaseView.TemplateResult(new QuaseView.Template([[\\"prop\\", 0, \\"someProp\\", [\\"\\", \\"\\"]], [\\"prop\\", 0, \\"a-nother\\", [\\"\\", \\"\\"]], [\\"prop\\", 0, \\"multiParts\\", [\\"\\", \\" \\", \\"\\"]], [\\"prop\\", 0, \\"\\\\uD83D\\\\uDC4D\\", [\\"\\", \\"\\"]], [\\"prop\\", 0, \\"(a)\\", [\\"\\", \\"\\"]], [\\"prop\\", 0, \\"[a]\\", [\\"\\", \\"\\"]], [\\"attr\\", 0, \\"a\\", [\\"\\", \\"\\"]], [\\"node\\", 2], [\\"prop\\", 3, \\"aThing\\", [\\"\\", \\"\\"]], [\\"node\\", 4]], \\"\\\\n    <div>\\\\n      <p><!--{{}}--></p>\\\\n      <div><!--{{}}--></div>\\\\n    </div>\\"), [

  1,
  2,
  3, 4,
  5,
  6,
  7,
  8,
  9,
  10, [1, new QuaseView.TemplateResult(new QuaseView.Template([[\\"node\\", 1]], \\"<span><!--{{}}--></span>\\"), [2]), 3]]);

};"
`;

exports[`compile and runtime test 0 2`] = `
"
    <div a=\\"8\\">
      <p>9</p>
      <div>1<span>2</span>3<!----></div>
    </div>"
`;
